lib/rules/color-contrast-matches.js

```diff
@@ -43,109 +61,74 @@ function colorContrastMatches(node, virtualNode) {
 				return false;
 			}
 		}
+		// Match all form fields, regardless of if they have text
+		return true;
 	}
-
-	if (nodeName === 'INPUT') {
-		return (
-			['hidden', 'range', 'color', 'checkbox', 'radio', 'image'].indexOf(
-				nodeType
-			) === -1 && !node.disabled
-		);
-	}
-
-	if (nodeName === 'SELECT') {
-		return !!node.options.length && !node.disabled;
-	}
-
-	if (nodeName === 'TEXTAREA') {
-		return !node.disabled;
-	}
-
-	if (nodeName === 'OPTION') {
-		return false;
-	}
-
-	if (
-		(nodeName === 'BUTTON' && node.disabled) ||
-		findUpVirtual(virtualNode, 'button[disabled]')
-	) {
-		return false;
-	}
-
-	if (
-		(nodeName === 'FIELDSET' && node.disabled) ||
-		findUpVirtual(virtualNode, 'fieldset[disabled]')
-	) {
-		return false;
-	}
-
-	// check if the element is a label or label descendant for a disabled control
 	const nodeParentLabel = findUpVirtual(virtualNode, 'label');
-	if (nodeName === 'LABEL' || nodeParentLabel) {
-		let relevantNode = node;
-		let relevantVirtualNode = virtualNode;
-
-		if (nodeParentLabel) {
-			relevantNode = nodeParentLabel;
-			// we need an input candidate from a parent to account for label children
-			relevantVirtualNode = getNodeFromTree(nodeParentLabel);
-		}
-		// explicit label of disabled input
-		const doc = getRootNode(relevantNode);
-		let candidate =
-			relevantNode.htmlFor && doc.getElementById(relevantNode.htmlFor);
-		const candidateVirtualNode = getNodeFromTree(candidate);
-
-		if (
-			candidate &&
-			(candidate.disabled ||
-				candidate.getAttribute('aria-disabled') === 'true' ||
-				findUpVirtual(candidateVirtualNode, '[aria-disabled="true"]'))
-		) {
+	if (nodeName === 'label' || nodeParentLabel) {
+		let labelNode = nodeParentLabel || node;
+		let labelVirtual = nodeParentLabel
+			? getNodeFromTree(nodeParentLabel)
+			: virtualNode;
+
+		// explicit label of disabled control
+		const doc = getRootNode(labelNode);
+		const explicitControl = doc.getElementById(labelNode.htmlFor || '');
+		const explicitControlVirtual =
+			explicitControl && getNodeFromTree(explicitControl);
+
+		if (explicitControlVirtual && isDisabled(explicitControlVirtual)) {
 			return false;
 		}
-
-		candidate = querySelectorAll(
-			relevantVirtualNode,
+		// implicit label of disabled control
+		const query =
```

Comment:
If we have an explicit control, do we need to check for an implicit one as well if the explicit control is not disabled?

Reply:
I don't think so. If an elm has an explicit label, the implicit one will not be applied.

Reply:
The way this `if` statement above is written, if the element has an explicit label and it's not disabled, it will continue to look at the implicit label. Based on your answer we would get into a state where if the explicit label is not disable but the implicit one is, we would ignore the element.

Reply:
Thought about this, I think fixing this is well outside of the scope of this PR. I'll open an issue for it instead.